File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.ts
function shipWithinDays(weights: number[], days: number): number {
    function getDays(capacity) {
        let D: number = 1
        let total: number = 0

        for (const weight of weights) {
            total += weight
            if (total > capacity) {
                total = weight
                D++
            }
        }
        return D
    }

    let left: number = Math.max(...weights)
    let right: number = weights.reduce((acc, cur) => acc + cur, 0)

    while (left < right) {
        const mid: number = Math.floor((left + right) / 2)
        const D: number = getDays(mid)
        if (D > days) left = mid + 1
        else right = mid
    }
    return right
};

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        low, high = max(weights), sum(weights)

        def can_ship(capacity: int) -> bool:
            total, days_needed = 0, 1

            for weight in weights:
                if total + weight > capacity:
                    days_needed += 1
                    total = 0
                total += weight
            return days_needed <= days

        while low < high:
            mid = (low + high) // 2
            if can_ship(mid):
                high = mid
            else: 
                low = mid + 1
        
        return low